File: lee/src/main/java/org/example/lee/题目/贪心/c_监控二叉树.java
package org.example.lee.题目.贪心;

import org.example.lee.model.TreeNode;
import org.example.lee.util.LeeUtil;

public class c_监控二叉树 {
    public int key = 0;
    public int minCameraCover(TreeNode root) {
       //根节点返回0的补偿
       if(tool(root)==0){
            key++;
       }
        return key;
    }

     int tool(TreeNode node){
        //0 我需要观测
        //1 我将成为摄像头观测我的子节点
        //2 我被观测了
        int tool = -1,tool1 = -1;
        if(node.left!=null) {
            tool = tool(node.left);
        }
        if(node.right!=null) {
            tool1 = tool(node.right);
        }
        if((tool == -1 ||tool == 2) && (tool1 == -1 ||tool1 == 2)){
            //自己是叶子节点 或者 子节点都被观测
            //我需要观测
            return 0;
        }
        if(tool == 0 || tool1 == 0){
            //有无观测节点 我要成为摄像头
            key++;
            //1 我是摄像头
            return 1;
        }
        if(tool == 1 || tool1 == 1){
            //2 我被观测了
            return 2;
        }
        return -1;
    }
}

File: lee/src/main/java/org/example/lee/题目/贪心/c_监控二叉树.go
package 贪心

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

var key int

func minCameraCover(root *TreeNode) int {
	key = 0
	i := tool(root)
	if i == 2 {
		key++
	}
	return key
}

func tool(root *TreeNode) int {
	//0 是监控
	//1 被监控
	//2 没被监控
	if root == nil {
		return 1
	}
	left := tool(root.Left)
	right := tool(root.Right)
	if left == 2 || right == 2 {
		key++
		return 0
	}
	if left == 1 && right == 1 {
		return 2
	}
	return 1
}
